### PR TITLE
Add check for verifying where crates are sourced from

### LINKED ISSUE #79: Add support for verify source of crate (git, registry)
Recently saw this article about a potential exploit surface in `npm` that should apply to Cargo and its lockfiles also: [Why npm lockfiles can be a security blindspot for injecting malicious modules](https://snyk.io/blog/why-npm-lockfiles-can-be-a-security-blindspot-for-injecting-malicious-modules/).

In Cargo for a binary which has a checked in `Cargo.lock` one can also edit it and change a crates.io dependency to a git dependency, our re-route dependencies in other ways to potentially less trusted sources.

I've started adding a new validation step specifically for this that enables only using explicitly allowed crate registries and git repositories.

### PR DESCRIPTION
**Pull request by repi**
# `cargo check sources`

## Overview & motivation

This enables one to only allow crates from specific registries (like crates.io, or custom ones), or allow specific git repositories (or none), or disallow all registries and repositories to only rely on file dependencies in the local repo (vendoring).

Inspired by the blog post: [Why npm lockfiles can be a security blindspot for injecting malicious modules](https://snyk.io/blog/why-npm-lockfiles-can-be-a-security-blindspot-for-injecting-malicious-modules/).

Fix #79.

## Example output

For cargo deny itself and without allowing any sources one gets the following output:

image

## TODO / design discussions

- [ ] Add tests to verify functionality
- [x] Cut: Deny by default instead?  
- [x] Keep crates.io as default allowed?
- [x] Later: Add wildcard URL searches?
- [x] Later: Highlight exact source span?

### Deny by default instead?

Right now this by default allows unknown registries and unknown git repositories, to not break any existing configurations. E.g:

```toml
[sources]
unknown-registry = "allow"
unknown-git = "allow"
```

Think it is the right choice so that a user have to explicitly enable the functionality. Though on the other hand it can be nice to be able to run `cargo deny check sources` on a new repo and actually get warnings on git dependencies or custom repositories. What do you think @Jake-Shadle ?

Adding a custom registry is quite uncommon and requires explicitly adding it in `.cargo/config` and pointing dependencies to it, it is typically less of a concern. Git dependencies on the other hand are more common.

Initially I didn't have crates.io as default allowed and then every single crate generated a warning which motivated this default. But now it would only be git dependencies or custom registries that would generate warnings.

### Keep crates.io as default allowed?

Right now one can do this to allow crates.io but deny any other registries:

```toml
[sources]
unknown-registry = "deny"
```

This works because we have a default value for `allow-registry` that points to crates.io:

```toml
allow-registry = [
    "https://github.com/rust-lang/crates.io-index"
]
```

I think I prefer this even though it is less explicit because I like being able to deny unknown registries in most configs but not having to specify crates.io everywhere. But one could argue that it would be more explicit and cleaner to not have a special case for crates.io. But it is just a default value and if one wants to be fully explicit one should specify all parameters and not rely on default values. Like this:

```toml
# deny all registries and git repositories
[sources]
unknown-registry = "deny"
unknown-git = "deny"
allow-registry = []
allow-git = []
```

### Add wildcard URL searches?

Using this in our internal `ark` monorepo currently requires the current config as we are using quite a few git forks patched in:

```toml
[sources]
unknown-registry = "deny"
unknown-git = "deny"
allow-git = [
    "https://github.com/EmbarkStudios/cpal",
    "https://github.com/EmbarkStudios/nfd-rs",
    "https://github.com/EmbarkStudios/winit",
    "https://github.com/EmbarkStudios/smush-rs.git",
    "https://github.com/EmbarkStudios/shaderc-rs.git",
    "https://github.com/EmbarkStudios/sentry-rust.git",

    "https://github.com/seanmonstar/reqwest.git",
]
```

It would be quite neat to support some type of URL wildcard matching as we trust ourselves, so one can specify it something like this instead:

```toml
[sources]
unknown-registry = "deny"
unknown-git = "deny"
allow-git = [
    # allow our own git forks
    "https://github.com/EmbarkStudios/",

    "https://github.com/seanmonstar/reqwest.git",
]
```

But could add that as a separate improvement afterwards

### Highlight exact source span?

It would would be nice in the output to highlight only the exact URL part of the source in the warnings. 

So instead of this:

```
error: detected crate git source "https://github.com/RustSec/rustsec-crate.git" not specifically allowed

    ┌── C:\git\embark\cargo-deny\Cargo.lock:89:1 ───
    │
 89 │ rustsec 0.16.0 git+https://github.com/RustSec/rustsec-crate.git?rev=aaba369#aaba369bebc4fcfb9133b1379bcf430b707188a2
    │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source
```

One can remove the source name in the error message and only rely on the exact highlighting:

```
error: detected crate git source not specifically allowed

    ┌── C:\git\embark\cargo-deny\Cargo.lock:89:1 ───
    │
 89 │ rustsec 0.16.0 git+https://github.com/RustSec/rustsec-crate.git?rev=aaba369#aaba369bebc4fcfb9133b1379bcf430b707188a2
    │                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ source    
```

This could also be a later minor improvement, not sure how to do it right now.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -349,6 +351,76 @@ The threshold for security vulnerabilities to be turned into notes instead of of
...
+
+#### The `unknown-git` field (optional)
+
+Determines what happens when a crate from a git repository not in the allow list is encountered.
+
+* `deny` - Will emit an error with the URL of the repository, and fail the check.
+* `warn` - Prints a warning for each crate, but does not fail the check.
+* `allow` (default) - Prints a note for each crate, but does not fail the check.
+
+#### The `allow_registry` field (optional)
```
Comment (Jake-Shadle) [inline README.md]:
```suggestion
#### The `allow-registry` field (optional)
```

File: README.md [hunk 2]
```diff
@@ -349,6 +351,76 @@ The threshold for security vulnerabilities to be turned into notes instead of of
...
+To not allow any crates registries, set to empty:
+
+```toml
+[sources]
+unknown-registry = "deny"
+allow-registry = []
+```
+
+
+#### The `allow_git` field
```
Comment (Jake-Shadle) [inline README.md]:
```suggestion
#### The `allow-git` field
```

File: src/sources/cfg.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    #[serde(default = "crate::lint_warn")]
+    pub unknown_git: LintLevel,
+
+    //    #[serde(default = "crate::lint_warn")]
+    //    pub missing_git_revision: LintLevel,
+    #[serde(default = "default_allow_registry")]
+    pub allow_registry: Vec<String>,
+
+    #[serde(default)]
+    pub allow_git: Vec<String>,
```
Comment (Jake-Shadle) [inline src/sources/cfg.rs]: I wonder if this should be able to just do allowances by user/org as well (at least for github sources), eg, 
```toml
allow-git = ["https://github.com/bytecodealliance"]
```

Comment (Jake-Shadle) [reply src/sources/cfg.rs]: Then again, generally you don't have so many git sources that that's a big deal.

Comment (repi) [reply src/sources/cfg.rs]: Yes  I'm thinking so as well, was first thinking of it as URL wildcard matching (see "Add wildcard URL searches?" section). But could potentially be additional config option here also, though implementation would be the same. 

Though the concept of an org is a GitHub concept in their URL, not a git one in general. So doing filtering/pattern matching on the git repo URL may be the cleanest.

An easy option would be to just check that the URL matches the length of it, then one can have a shorter URL with just the org name in the allow list

**Comment by Jake-Shadle** [comment]
So bullet points

- Deny by default instead? I think warn is fine here, though adding #61 would allow people to more easily turn the warnings to errors in eg CI
- Keep crates.io as default allowed? Yes, and if they modify `allow-registries` they will already have to readd it explicitly if they want it still.
- Add wildcard URL searches? For Github sources, this could be done for eg user or org, but considering how few git sources there are, this just seems like overkill. That being said, a new subcommand could be added, similar to `cargo deny init` that could auto fill all of the current sources into the allow list.
- Highlight exact source span? Would be nice, but each crate is on its own line and the components are fairly self describing, so not a huge deal.

**Comment by Jake-Shadle** [comment]
Oh yes, and you probably want to modify the `resources/template.toml`

**Comment by repi** [comment]
Ah didn't see that one. Will add.

**Comment by repi** [comment]
And thanks for the feedback! 

Currently for unknown registries and git repos it has lint level "allow", think I'll change that to "warn". Will only by default then warn about git repos that one uses but that should be fine, and will prompt users that see it to look into this new functionality and configure it how they want it

**Pull request merged**